morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+            morphology_folder = neurondb.parent.resolve()
+
+        if neurondb.suffix.lower() == '.dat':
+            columns = ['name', 'layer', 'mtype']
+            obj.df = pd.read_csv(neurondb, sep=r'\s+', names=columns, usecols=range(len(columns)))
+
+            fulltypes = obj.df.mtype.str.split(':', n=1, expand=True).fillna('')
+            if len(fulltypes.columns) > 1:
+                obj.df['mtype_no_subtype'] = fulltypes[0]
+                obj.df['msubtype'] = fulltypes[1]
+            else:
+                obj.df['mtype_no_subtype'] = obj.df.mtype
+                obj.df['msubtype'] = ''
+            obj.df['label'] = label
+            for missing_col in set(COLUMNS) - set(obj.df.columns):
+                obj.df[missing_col] = None
+            obj.df.layer = obj.df.layer.astype('str')
+            obj.df['path'] = obj.df.apply(lambda row: find_morph(
+                morphology_folder, row['name']), axis=1)
+            obj.df = obj.df.reindex(columns=COLUMNS)
+            for key in BOOLEAN_REPAIR_ATTRS:
+                obj.df[key] = True
+            obj.df['axon_inputs'] = [[] for _ in range(len(obj.df))]
+
+            return obj
+
+        with neurondb.open() as fd:
+            content = fd.read()
+            neurondb = xmltodict.parse(content)
+
+        morphologies = neurondb['neurondb']['listing']['morphology']
+
+        # Case where there is a single <morphology></morphology> tag in the XML
+        if not isinstance(morphologies, list):
+            morphologies = [morphologies]
+        morphologies = filter(None, morphologies)
+        morphologies = list(map(MorphInfo._from_xml, morphologies))  # noqa, pylint: disable=protected-access
+
+        for morph in morphologies:
+            morph.label = label
+            morph.path = find_morph(morphology_folder, morph.name)
+
+        obj.df = _create_dataframe(morphologies)
+        obj.df = obj.df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS})
+        return obj
+
+    @classmethod
+    def from_folder(cls,
+                    morphology_folder: Path,
+                    mtypes: Dict[str, str],
+                    label: str = 'default'):
+        '''Factory method to create a MorphDB object from a folder containing morphologies
+
+        Args:
+            morphology_folder: a folder containing morphologies
+            mtype: a dictionary of where key is a morphology name and value its mtype
+            label: (optional) a group label to be used to identify the morphlogies from this folder
+        '''
+
+        obj = cls()
+        obj += (MorphInfo(path.stem, mtypes[path.stem], label=label)
+                for path in iter_morphology_files(morphology_folder))
+        return obj
+
+    def write(self, output_path: Path):
+        '''Write the neurondb file to XML or DAT format'''
+        ext = output_path.suffix.lower()
+        if ext == '.dat':
+            self.df[['name', 'layer', 'mtype']].to_csv(output_path, sep=' ', header=False,
```

Comment:
we should allow to dump all the df, no?

Reply:
I am not sure creating a new format is a good idea. If we need this, we'll do this in another review.

Reply:
Yes, we can always to `mydb.df.to_csv('all_data.csv')` for now